packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -2696,18 +2701,26 @@ export class StripeHelper extends StripeHelperBase {
       );
     }
 
-    const previousAttributes = eventData.previous_attributes;
+    // Stripe only sends fields that have changed in their previous_attributes field
+    // Additionally, previous_attributes is a generic field that has no proper typings
+    // and is used in a flexible manner.
+    const previousAttributes = eventData.previous_attributes as any;
+    const planOldDiff = (previousAttributes as any)
+      .plan as Partial<Stripe.Plan> | null;
+    const planOld: Stripe.Plan | null = planOldDiff
+      ? {
+          ...planNew,
```

Comment:
From what I can tell, there is no risk here. The previous_attributes from stripe contains all fields that have changed between the old plan and the new plan according to their documentation.

Unfortunately in cases where the two plans have the same interval_count, previous_attributes will not contain interval_count, and potentially other critical fields down the line. In the old code, previous_attributes.plan was treated as a proper Stripe.Plan which is incorrect and will often be missing critical information (thus the nullish coalesce to the newPlan details that was taking place in the old code). Correctly typed, previous_attributes.plan is a Partial<Stripe.Plan>.

I see this as the best way of getting a proper Stripe.Plan representing the old plan, since I don't see a way to get a Stripe.Plan by query without hitting the Stripe plan API itself. We only have utils to fetch AbbrevPlan, which doesn't contain all of the necessary details we use later in subsequent utilities.

Reply:
Ah OK sweet that makes sense. Yup, I agree, I think it's safe to do this.